File: tests/hibernate5/hibernate5-sync-common/src/main/java/example/sync/PetRepository.java
/*
 * Copyright 2017-2020 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package example.sync;

import example.domain.IPet;
import io.micronaut.transaction.TransactionDefinition;
import io.micronaut.transaction.annotation.TransactionalAdvice;
import jakarta.inject.Singleton;
import org.hibernate.Hibernate;

import javax.persistence.EntityManager;
import java.util.Collection;
import java.util.Optional;

@Singleton
class PetRepository implements IPetRepository {

    private final EntityManager entityManager;

    PetRepository(EntityManager entityManager) {
        this.entityManager = entityManager;
    }

    @Override
    public IPet create() {
        return new Pet();
    }

    @TransactionalAdvice(propagation = TransactionDefinition.Propagation.MANDATORY)
    @Override
    public void save(IPet pet) {
        entityManager.persist(pet);
    }

    @TransactionalAdvice(propagation = TransactionDefinition.Propagation.MANDATORY)
    @Override
    public void delete(IPet pet) {
        entityManager.remove(pet);
    }

    @Override
    public Collection<Pet> findAll() {
        return entityManager.createQuery("from Pet", Pet.class).getResultList();
    }

    @Override
    public Optional<IPet> findByName(String name) {
        return entityManager.createQuery("from Pet where name = :name", Pet.class)
                .setParameter("name", name)
                .getResultList()
                .stream()
                .findFirst()
                .map(pet -> {
                    if (Hibernate.isInitialized(pet.getOwner())) {
                        throw new IllegalStateException("Expected not initialized owner ref");
                    }
                    pet.getOwner().getName();
                    if (!Hibernate.isInitialized(pet.getOwner())) {
                        throw new IllegalStateException("Expected initialized owner ref");
                    }
                    return pet;
                });
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/hibernate/proxies/PetRepository.kt
package io.micronaut.docs.hibernate.proxies

import io.micronaut.transaction.annotation.Transactional
import jakarta.inject.Singleton
import jakarta.persistence.EntityManager
import jakarta.persistence.PersistenceContext
import org.hibernate.Hibernate

@Singleton
open class PetRepository {

    @PersistenceContext
    lateinit var entityManager: EntityManager

    @Transactional
    open fun save(petName: String, ownerName: String): Pet {
        val owner = Owner()
        owner.name = ownerName
        entityManager.persist(owner)
        val pet = Pet()
        pet.name = petName
        pet.owner = owner
        entityManager.persist(pet)
        return pet
    }

    @Transactional(readOnly = true)
    open fun isOwnerLazilyLoaded(petId: Long): Boolean {
        val pet = entityManager.find(Pet::class.java, petId)
        val owner = pet.owner!!
        val initialized = Hibernate.isInitialized(owner)
        owner.name
        return !initialized && Hibernate.isInitialized(owner)
    }

    @Transactional(readOnly = true)
    open fun ownerClass(petId: Long): Class<*> = entityManager.find(Pet::class.java, petId).owner!!.javaClass
}
